appkit/src/test/scala/org/ergoplatform/appkit/BabelFeeSpec.scala

```diff
@@ -0,0 +1,168 @@
+package org.ergoplatform.appkit
+
+import org.ergoplatform.appkit.babelfee.{BabelFeeBoxContract, BabelFeeBoxState, BabelFeeOperations}
+import org.scalatest.{Matchers, PropSpec}
+import org.scalatestplus.scalacheck.ScalaCheckDrivenPropertyChecks
+
+import java.util
+import java.util.Arrays
+
+class BabelFeeSpec extends PropSpec with Matchers with ScalaCheckDrivenPropertyChecks
+  with HttpClientTesting
+  with AppkitTestingCommon {
+
+  private val script =
+    """{
+      |
+      |    // ===== Contract Information ===== //
+      |    // Name: EIP-0031 Babel Fees Contract
+      |    // Description: Contract guarding the babel fee box, checking if valid output babel box was recreated and the token exchange was valid.
+      |    // Version: 1.0.0
+      |
+      |    // ===== Relevant Variables ===== //
+      |    val babelFeeBoxCreator: SigmaProp = SELF.R4[SigmaProp].get
+      |    val ergPricePerToken: Long = SELF.R5[Long].get
+      |    val tokenId: Coll[Byte] = _tokenId
+      |    val recreatedBabelBoxIndex: Option[Int] = getVar[Int](0)
+      |
+      |    // ===== Perform Babel Fee Swap ===== //
+      |    if (recreatedBabelBoxIndex.isDefined) {
+      |
+      |        // Check conditions for a valid babel fee swap
+      |        val validBabelFeeSwap: Boolean = {
+      |
+      |            // Output babel fee box
+      |            val recreatedBabelBox: Box = OUTPUTS(recreatedBabelBoxIndex.get)
+      |
+      |            // Check that the babel fee box is recreated correctly
+      |            val validBabelFeeBoxRecreation: Boolean =
+      |
+      |                allOf(Coll(
+      |                    (recreatedBabelBox.propositionBytes == SELF.propositionBytes),
+      |                    (recreatedBabelBox.tokens(0)._1 == tokenId),
+      |                    (recreatedBabelBox.R4[SigmaProp].get == babelFeeBoxCreator),
+      |                    (recreatedBabelBox.R5[Long].get == ergPricePerToken),
+      |                    (recreatedBabelBox.R6[Coll[Byte]].get == SELF.id)
+      |                ))
+      |
+      |
+      |
+      |            // Check that the user's token was exchanged correctly
+      |            val validBabelFeeExchange: Boolean = {
+      |
+      |                val nanoErgsDifference: Long = SELF.value - recreatedBabelBox.value
+      |                val babelTokensBefore: Long = if (SELF.tokens.size > 0) SELF.tokens(0)._2 else 0L
+      |                val babelTokensDifference: Long = recreatedBabelBox.tokens(0)._2 - babelTokensBefore
+      |
+      |                allOf(Coll(
+      |                    (babelTokensDifference * ergPricePerToken >= nanoErgsDifference),
+      |                    (nanoErgsDifference >= 0)
```

Comment:
It seems recreatedBabelBox can be clone of SELF, is this ok?

Reply:
It does no real harm and can be used to refresh the box to avoid storage fee